Add custom tag like Blade without need to define the module href path

thewebartisan7 opened the pull request:
I hope you will like this feature, which allow to use Blade-like syntax for load modules.

See blow code example:

```html
  <x-my-module>
    Content my-module...
  </x-my-module>

<!-- Use dot for find module inside folder -->
  <x-forms.button class="bg-primary">
    Content button.html
  </x-forms.button>

<!-- Use namespace prefix for find module inside defined path -->
  <x-theme-dark::my-dark-module class="text-dark">
    Content my-dark-module.html
  </x-theme-dark::my-dark-module>
```

All test has been passed and I added a few new.

I was not sure how to handle when module was not found, so for now I throw new error, but in case we can just return node as-is, or also add strict mode true/false for raise error or return node as-is.

I am also tempted to add multiple slots, like extends plugin has, but for now I keep this as simple as possible.

Let me know what do you think, I will be glad to adjust and fix any feedback you may have.

thewebartisan7 commented:
I would like to add one more thing to this PR which is also available in Blade, read here https://laravel.com/docs/9.x/blade#anonymous-index-components

Shortly, if you want to group your multiple module's files inside a folder, like for example:

/modules/accordion/index.html
/modules/accordion/item.html

Blade allow to use tag:

```html
<x-accordion>
    <x-accordion.item>
        ...
    </x-accordion.item>
</x-accordion>
```

So it retrieve index.html from subfolder without need to specify it.

cossssmin commented:
This is pretty awesome 👏 

@Scrum what do you think?

Scrum commented:
@thewebartisan7 yes, it looks great. have a couple of questions:
1. Obviously, there are nested folder sublevels for components, for example:
`rootComponentFolder->button->index.html` `rootComponentFolder->layout->footer->index.html`
2. Supplements the first question, most likely the theme is placed in the component folder, for example: `rootComponentFolder->button->theme-dark->index.html`

What do you think about this ?

thewebartisan7 commented:
I am glad to hear that you like it!

> Obviously, there are nested folder sublevels for components, for example:
> rootComponentFolder->button->index.html rootComponentFolder->layout->footer->index.html

Maybe I didn't get your question, but I don't see problem here.
Once you define one or multiple roots, then it first search inside this the exact match with root + tag name (including sublevel only when there are dots).
When not found any of this, then it search for the index.html at only first level, and not in sublevels, if dots are not in the tag name.

For example `rootComponentFolder->button->index.html`  would be:

```html
<x-button>my button</x-button>
<!-- or also: -->
<x-button.index>my button</x-button.index>
```

While the second case would be:

```html
<x-layout.footer>my footer</x-layout.footer>
<!-- or also: -->
<x-layout.footer.index>my footer</x-layout.footer.index>
```

> Supplements the first question, most likely the theme is placed in the component folder, for example: 
> rootComponentFolder->button->theme-dark->index.html

A namespace should be a top level folder. In case of a `theme-dark` namespace, you don't want to have multiple themes folder, but just one with all components. For example the structure would be:

- modules
  - theme-dark
    - button/index.html
    - input/index.html
    - label/index.html
  - theme-light
    - button/index.html
    - input/index.html
    - label/index.html

Without namespace you could also use:

```html
<x-theme-dark.button></x-theme-dark.button>
<!-- or -->
<x-theme-dark.button.index></x-theme-dark.button.index>
```

another theme:

```html
<x-theme-light.button></x-theme-light.button>
<!-- or -->
<x-theme-light.button.index></x-theme-light.button.index>
```

The good of namespace is that you can define, and change in future, the folder name without changing the code.
And it didn't prepend the root, it has their own root, so you could have your namespace folder anywhere, in a separated repo for example. Blade use them especially for package development. I could imagine an UI for Bootstrap or Tailwind that reside in another repository so you can install and use via namespace, with custom namespace.
So when you are using namespace, it will look only in that defined root, and will not take into account any other root that you have defined.

Please let me know if I am missing something.

thewebartisan7 commented:
I have refactor the code a bit for better readability, and also added fallback index.html when not found any other in defined roots or by namespace root.

I also notice that before I was always using options.root even with namespace, but now this is fixed.
Since root was always taken into account in `readFile` method apart if href is relative, I have so used this by setting `options.from` when using namespace root.

I will review again in next days and write some more tests.

Let me know any feedback you may have.

Thanks

Scrum commented:
In general, everything sounds very reasonable, there was a small question. In the case of hoards of themings, is there a fallback if there is no component in the theming?

Scrum commented on index.js:
```diff
@@ -115,19 +125,115 @@ function parse(options) {
+    throw new Error(`Unknown module namespace ${namespace}.`);
+  }
+
+  if (!fs.existsSync(`${customTagRoot}/${customTagFile}`)) {
```
Here it is better to use `path.join` because For windows OS this will not work.

Scrum commented on index.js:
```diff
@@ -115,19 +125,115 @@ function parse(options) {
+    .replace(options.customTagPrefix, '')
+    .split('.')
+    .join('/')
+    .concat('.', options.customTagExtension);
```
all these `.`, `/` values must be converted to a constant and consider for `\` OS features

thewebartisan7 replied:
I was not sure if add folder separator in options so for now is as constant. I could not imagine another separator for folder inside tag name. 

I added only namespace separator in options.

Scrum commented on index.js:
```diff
@@ -115,19 +125,115 @@ function parse(options) {
+  // Find module by defined namespace in options.customTagNamespaces
+  //  or by defined roots in options.customTagRoot
+  //  and set the returned path
+  node.attrs[options.attribute] = tag.includes('::') ?
```
`::` must be converted to a constant or defined in the options

thewebartisan7 commented:
> In general, everything sounds very reasonable, there was a small question. In the case of hoards of themings, is there a fallback if there is no component in the theming?

It's interesting idea! Something like Blade also manage. Let me think about this.

thewebartisan7 commented:
I added fallback for namespace root, and even a custom / override root to search first, which can be useful if you want to override a specific namespace module, for example when this is a third party module.

So first you define your `customTagNamespaces` options as an array this time like below:

```
plugin({
  customTagNamespaces:  [
   {
    name: 'theme-dark',
    root: './theme-light/modules/',
    custom: './theme-light-custom/modules/',
    fallback: './theme-default/modules/'
   }
  ]
)
```

`custom` and `fallback` are optional, and when not provided, will be used only the root and fail when not found.

When provided the `custom` then first of all the `findModuleByNamespace` will search inside the `custom` root, and when not found will check in `root` and as last, when defined, in the `fallback`.

We can also use the global option applied to all namespace `customTagNamespaceFallback` which by default is `false`. When is `true` then we will search inside all defined `customTagRoots`.

I think that in this way we are covering everything regarding override and fallback.

Optionally we could allow that `customTagNamespaceFallback` is a string as root separated from main root?

I will check again everything this weekend and also add all this info in readme.

Let me know what do you think.

Scrum commented:
Do not consider me a bore, but why did you decide to implement this feature in this package? Is there any desire to bring this great feature into a separate package?

cossssmin commented:
> Is there any desire to bring this great feature into a separate package?

In my opinion there would be a lot of duplication and confusion if this feature were to live in a separate package. I am all for it to be part of `posthtml-modules`.

thewebartisan7 commented:
@Scrum  I understand your concern and I see the philosophy of PostHTML is to offer small plugins that do just one thing, and I agree with this.

I also tough about making this a separated plugin, but that would be yet another similar plugins as @cossssmin  said.

Since I start use PostHTML I found that posthtml-include, posthtml-extends and posthtml-modules could be combine into one, or almost posthtml-extends and posthtml-modules. 

Because I like the blocks of posthtml-extends, and the attributes as local of posthtml-modules.

In my opinion should be just one, with blocks/slots, attributes as local and the custom tag.

But in case you want to keep this separated, I could think a way to make this available to both extends and modules. I am already using both in my project, so it's fine for me.

Because I was also thinking how to do the same in extends plugin without code duplication.

Although initially I didn't think that will grow so much quickly with feature.

I am open to any suggestions to handle this.

cossssmin commented:
@thewebartisan7 I'd like to get in touch, can you please ping me either [@cossssmin](https://twitter.com/cossssmin) on Twitter or cosmin at mailviews com? 

Thanks!

thewebartisan7 commented:
I try to create a separated plugin that can work with `posthtml-modules` and `posthtml-extends`, and in theory also with `posthtml-include` or any other similar plugin that has an attribute with a path.

This way it follow the PostHTML principle of "do one thing, and do it well".

I need to create a readme, but you can already see most of test with 100% coverage, however it still need to test when different options are set in the plugin and/or in others above mentioned plugins, especially when a custom tag is replaced in modules or extends.

I want to review the part that set path when namespace is used, because extends src require a relative path, while modules require absolute path, for this reason I added a boolean `options.absolute` to use **only** when the extends or modules tag name are custom, because otherwise it's auto set to true when tag is modules, see https://github.com/thewebartisan7/posthtml-custom-tag/blob/main/src/index.js#L176. For example a way would be to check the function name that is set for extend plugin (`require('posthtml-extend').name`), however is anynomous function in module, and could be also for other plugin. As said, I will review this.

There is two way of using it, by applying the plugin to tree before the modules and extends plugin, or by passing the modules and extends plugin via options.

Example applying the plugin before the modules:

```js
  const html = await posthtml([
    customTag({root: './', roots: ['test/templates/modules/'], attribute: 'href', replaceTagNameWith: 'module'}),
    posthtmlModules({root: './'})
  ])
    .process(actual)
    .then(result => clean(result.html));
```

Example passing module plugin via options:

```js
  const html = await posthtml([
    customTag({root: './', roots: 'test/templates/modules/', attribute: 'href', replaceTagNameWith: 'module', plugins: [posthtmlModules({root: './'})]})
  ])
    .process(actual)
    .then(result => clean(result.html));
```

In second way we could also pass options from one to another and vice-versa, but for now this is not yet possible.

Another thing that I want to review is to allow to use the same prefix for both plugin, because we could simple return node as-is when it's not found the path, so the next plugin can be called two times one for module and one for extend, example:

```js
  const html = await posthtml([
    customTag({root: './', roots: 'test/templates/modules/', attribute: 'href', replaceTagNameWith: 'module', plugins: [posthtmlModules({root: './'})]}),
customTag({root: './', roots: 'test/templates/layouts/', attribute: 'src', replaceTagNameWith: 'extends'}),
    posthtmlExtend({root: './'})
  ])
    .process(actual)
    .then(result => clean(result.html));
```

This is not yet tested but with new options.strict we can trigger error when not found or just return the node so the second time it will be found.

You can find here: https://github.com/thewebartisan7/posthtml-custom-tag

In case we go with this solution, we can move the repo under posthtml organization.

I am also checking for a custom plugin that has extends slots, attributes as locals and this set custom tag, just for learning more and understand PostHTML better.

In case you are also interested in this, I would be glad to continue working on something with you.

@cossssmin sure I will get in touch with you via email.

thewebartisan7 commented:
Using extends and components an example of usage would be:

```html
<x-layout><!-- Extends layout.html -->

<slot name="header">
<x-header>My page</x-header><!-- Module header.html  -->
</slot name="header">

<slot name="content>
<x-form action="process.html" method="POST"><!-- Module form.html  -->
  <x-select options="{{ options }}"></x-select><!-- Module select.html  -->
  <x-input value="{{ value }}"></x-input><!-- Module input.html  -->
  <x-button>Submit</x-button><!-- Module button.html  -->
</x-form>
</slot>
</x-layout>
```

thewebartisan7 commented:
@Scrum  Please see here https://github.com/posthtml/posthtml/discussions/389

I did simplify the code for retrieve the path of module, see https://github.com/thewebartisan7/posthtml-components/blob/main/src/find-path.js

In case you are interested I can also update this PR, or we can just close and continue to discuss about new plugin in the discussion.

As I wrote in the discussion tab, this module could be improved how the content is added, because I found issue with nesting components inside component like shown in example. This happen if your component included other components and the <content> tag is inside nested component only. An infinite loop happen in this case. 

Also I found elegantly written this plugin, but recursively function + promise on an unknown tree, it's hard to understand. 

Without locals, see this version:

```js
'use strict';

const fs = require('fs');
const path = require('path');
const {parser} = require('posthtml-parser');
const {match} = require('posthtml/lib/api');

/**
 * @param  {Object} options Plugin options
 * @return {Object} PostHTML tree
 */
function processTree(options) {
  return function (tree) {
    match.call(tree, {tag: options.tag}, componentNode => {
      const filePath = path.join(options.root, componentNode.attrs[options.attribute]);
      const componentContent = fs.readFileSync(filePath, 'utf8');
      const componentTree = parser(componentContent);
      const content = match.call(componentTree, {tag: 'content'}, contentNode => {
        return componentNode.content || '';
      });
      componentNode.tag = false;
      componentNode.content = content;
      return componentNode;
    });
    return tree;
  };
}

module.exports = (options = {}) => tree => {
  options.locals = options.locals || {};
  options.tag = options.tag || 'module';
  options.plugins = options.plugins || [];
  options.attribute = options.attribute || 'href';
  options.root = path.resolve(options.root || './');

  tree = processTree(options)(tree);

  return tree;
};
```

Looking for to hear from you.

Scrum commented:
@thewebartisan7 Sorry, I fell out a little in time, please guide me on this task.

I roughly remembered and the idea seems very interesting to me, I can merge it if everything has already been done.

thewebartisan7 commented:
> @thewebartisan7 Sorry, I fell out a little in time, please guide me on this task.
> 
> I roughly remembered and the idea seems very interesting to me, I can merge it if everything has already been done.

No problem. I did as you suggest [here](https://github.com/posthtml/posthtml-modules/pull/94#issuecomment-1271192399), means I did a [separated plugin ](https://github.com/thewebartisan7/posthtml-components) for this.

Firstly I create a plugin that handle only custom tag, see https://github.com/thewebartisan7/posthtml-custom-tag, with the idea that could be used with posthtml-extends, posthtml-include, posthtml-modules and potentially with any other similar plugin. However I found several issues and then decided to create a new one that combine functionality of posthtml-extends (blocks/slots) and posthtml-modules (locals as attributes) and added some additional features.

You maybe remember that I announce it here https://github.com/posthtml/posthtml/discussions/389

I will add it to awesome-plugin list as you suggest [here](https://github.com/posthtml/posthtml/discussions/389#discussioncomment-4086185) as soon I release the first stable version.

At the moment it has not known bug, but I would like to solve two issue, first with [async plugin](https://github.com/thewebartisan7/posthtml-components/issues/21)  and this [second about html attributes](https://github.com/thewebartisan7/posthtml-components/issues/24).

I think that you can close this one, or if you wish to keep this feature also here in this plugin, then the current PR should works fine as I remember. The only issue you raised is the amount of code I added here, I suppose you want to keep it more simple. Am I right? In this case we could remove some feature like namespaces, since the basic code for convert custom tag to module that with path attributes is pretty simple.

Let me know how you wish to proceed. In case I can help you to do so.

Scrum commented:
@thewebartisan7 I think it's better to stabilize your plugin and then reuse it in the current plugin if needed.
